runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -265,34 +268,61 @@ MM_VLHGCAccessBarrier::jniGetPrimitiveArrayCritical(J9VMThread* vmThread, jarray
 		shouldCopy = true;
 	}
 
-	if (shouldCopy) {
-		GC_ArrayObjectModel* indexableObjectModel = &_extensions->indexableObjectModel;
-		I_32 sizeInElements = (I_32)indexableObjectModel->getSizeInElements(arrayObject);
-		UDATA sizeInBytes = indexableObjectModel->getDataSizeInBytes(arrayObject);
-		data = functions->jniArrayAllocateMemoryFromThread(vmThread, sizeInBytes);
-		if(NULL == data) {
-			functions->setNativeOutOfMemoryError(vmThread, 0, 0);	// better error message here?
-		} else {
-			indexableObjectModel->memcpyFromArray(data, arrayObject, 0, sizeInElements);
-			if(NULL != isCopy) {
-				*isCopy = JNI_TRUE;
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+	bool successDoubleMap = false;
+	if (shouldCopy && _extensions->indexableObjectModel.isDoubleMappingEnabled()) {
+		if (_extensions->indexableObjectModel.numArraylets(arrayObject) > 1) {
+			J9Object *firstLeafSlot = (J9Object *)((uintptr_t)_extensions->indexableObjectModel.getArrayoidPointer(arrayObject)[0]);
+			MM_HeapRegionDescriptorVLHGC *firstLeafRegionDescriptor = (MM_HeapRegionDescriptorVLHGC *)_extensions->heapRegionManager->tableDescriptorForAddress(firstLeafSlot);
+			data = firstLeafRegionDescriptor->_identifier.address;
+
+			if (NULL == data) {
+				/* Doublemap failed, but we still need to continue execution; therefore fallback to previous approach */
+				successDoubleMap = false;
+			} else {
+				successDoubleMap = true;
 			}
+		/* Corner case where there's only one arraylet leaf */
+		} else if (_extensions->indexableObjectModel.numArraylets(arrayObject) == 1 && _extensions->indexableObjectModel.getSizeInElements(arrayObject) > 0) {
+			/* Solo arraylet leaf is contiguous so we can simply return the data associated with it */
+			MM_JNICriticalRegion::enterCriticalRegion(vmThread, true);
+			Assert_MM_true(vmThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS);
+			data = (J9Object *)((uintptr_t)_extensions->indexableObjectModel.getArrayoidPointer(arrayObject)[0]);
+			successDoubleMap = true;
```

Comment:
based on this case (which effectively does not double map), would it make sense to call the flag something like contiguousDataAvailable?